scripts/hca-utils/hca_utils/utils.py

```diff
@@ -0,0 +1,332 @@
+import csv
+from functools import lru_cache
+import os
+from requests.exceptions import HTTPError
+from typing import Set
+import urllib.parse
+
+import google.auth
+from google.auth.transport.requests import AuthorizedSession
+from google.cloud import bigquery, storage
+from requests_cache.core import CachedSession
+
+
+class HcaUtils:
+    def __init__(self, environment: str, project: str, dataset: str):
+        self.environment = environment
+
+        if environment == "dev":
+            self.project = "broad-jade-dev-data"
+        else:
+            self.project = project
+
+        self.dataset = dataset
+
+        self.filename_template = f"sd-{project}-{dataset}-{{table}}.csv"
+
+        self.bigquery_client = bigquery.Client(project=self.project)
+
+        bucket_projects = {"prod": "mystical-slate-284720",
+                           "dev": "broad-dsp-monster-hca-dev"}
+        self.bucket_project = bucket_projects[environment]
+        self.bucket = f"broad-dsp-monster-hca-{environment}-staging-storage"
+
+        jade_urls = {"prod": "https://jade-terra.datarepo-prod.broadinstitute.org",
+                     "dev": "https://jade.datarepo-dev.broadinstitute.org"}
+        self.base_url = jade_urls[environment]
+
+        # this is always prod for bigquery, storage, etc.
+        gcp_creds, _ = google.auth.load_credentials_from_file("path/to/prod/credentials.json")
+        self.gcp_creds = gcp_creds
+
+        # this depends on the Jade env to interact with
+        creds_path = {"prod": "path/to/prod/credentials.json",
```

Comment:
I am not proud of this

Reply:
maybe a simple thing could be to just have environment variables point to these paths? Y'all would still need to go through the motions of having separate creds files and pointing to them though...

Reply:
the best solution here is to read the creds from vault IMO

Reply:
fwiw, I think google.auth already [supports](https://google-auth.readthedocs.io/en/latest/reference/google.auth.html?highlight=load_credentials_from_file#google.auth.load_credentials_from_file) an env var for this purpose (GOOGLE_APPLICATION_CREDENTIALS).

I'd be fine with an env var, I have no opinion about vault for this purpose. If you're going to stick with this approach, can the path be config'ed out or passed in via a cli arg?

Reply:
Since this is for a user (aka one of us) running it, I'm open to storing creds in vault but I would think that's better for a service account (at the very least, Monster hasn't stored personal creds in vault, but if we can/should then sure!)

Yup, that is the env var, but the issue is that we have more than one credentials file to use here. If we set the env var or let it use the default location of the file, and you have run `gcloud auth application-default login` and are logged in with your Broad account, then it'll work great for prod. But if you want to use the dev repo and you have a separate gmail account for that, then this doesn't work because:

- you need your Broad account's credentials for interacting with GCP
- you need your dummy dev account's credentials for interacting with the dev Repo

I suppose another solution here is to simply add our dev email accounts to our dev GCP projects maybe? The dev accounts can already access BigQuery because of Jade ACL setting stuff.

Thoughts @aherbst-broad @ThePletch ?

Reply:
hm. given that this is a task that we'll be running as part of production work, my instinct is that we shouldn't be running it with personal credentials in the first place (generally i think any work that's run on live data should be run with a SA that has the permissions it needs and no more, for both auditing purposes and purposes of self-protection, to make sure the script simply isn't allowed to do anything it's not intended to do).

open to discussion if people think that's overkill for this script or otherwise isn't a good idea.

Reply:
I disagree re: a shared service account:
1. What happens when someone leaves the org?
2. We get _less_ auditabilty with a shared acct since all ops will be attributed to the shared account, but potentially run by all devs

Either way, that doesn't help the nut of the problem (needing a dev acct for jade-dev and a prod account for gcp). How hard would be it be to add the dev accts to gcp? I think I like separating the prod from the dev credentials to minimize risk

Reply:
Okay so if we want to just address the issue of adding dev emails to dev GCP projects and then we don't have to do any of this file stuff (it'd just be your standard google application default credentials), we could create a user group for monster-dev emails and add them to our dev projects? Idk if that breaks any secops guidance